From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: glitch <glitch.g3431@gmail.com>
Date: Tue, 11 Jul 2023 17:40:07 -0500
Subject: [PATCH] Initial Quiltification

This includes:
- Changing the plugin marker (except in tests)
- Using a different set of caches
- Using the Quilt Loader system properties
- Using `quilt_installer.json`
- Renaming the shared service manager
- Automatically applying the Quilt Maven

diff --git a/build.gradle b/build.gradle
index a7ea089714500202f911ed736ddfb8fbc604ee3a..09543e2db653f6f426a4bb6c83aaa3dfebef3cd1 100644
--- a/build.gradle
+++ b/build.gradle
@@ -48,7 +48,7 @@ tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
 	}
 }
 
-group = 'net.fabricmc'
+group = 'org.quiltmc'
 def baseVersion = '1.8'
 
 def ENV = System.getenv()
@@ -273,10 +273,16 @@ codenarc {
 
 gradlePlugin {
 	plugins {
-		fabricLoom {
-			id = 'fabric-loom'
+		quiltLoom {
+			id = 'org.quiltmc.loom'
 			implementationClass = 'net.fabricmc.loom.bootstrap.LoomGradlePluginBootstrap'
 		}
+		if (!ENV.MAVEN_URL) {
+			fabricLoom {
+				id = 'fabric-loom'
+				implementationClass = 'net.fabricmc.loom.bootstrap.LoomGradlePluginBootstrap'
+			}
+		}
 	}
 }
 
@@ -329,8 +335,8 @@ publishing {
 
 			// Manually crate the plugin marker for snapshot versions
 			snapshotPlugin(MavenPublication) { publication ->
-				groupId 'fabric-loom'
-				artifactId 'fabric-loom.gradle.plugin'
+				groupId 'org.quiltmc.loom'
+				artifactId 'org.quiltmc.loom.gradle.plugin'
 				version baseVersion + '-SNAPSHOT'
 
 				pom.withXml({
@@ -340,9 +346,9 @@ publishing {
 					Node dependencies = root.appendChild(document.createElement('dependencies'))
 					Node dependency = dependencies.appendChild(document.createElement('dependency'))
 					Node groupId = dependency.appendChild(document.createElement('groupId'))
-					groupId.setTextContent('net.fabricmc')
+					groupId.setTextContent('org.quiltmc')
 					Node artifactId = dependency.appendChild(document.createElement('artifactId'))
-					artifactId.setTextContent('fabric-loom')
+					artifactId.setTextContent('loom')
 					Node version = dependency.appendChild(document.createElement('version'))
 					version.setTextContent(baseVersion + '-SNAPSHOT')
 				})
@@ -351,6 +357,7 @@ publishing {
 	}
 	repositories {
 		maven {
+			name = 'release'
 			if (ENV.MAVEN_URL) {
 				url ENV.MAVEN_URL
 				credentials {
@@ -359,6 +366,26 @@ publishing {
 				}
 			}
 		}
+		maven {
+			name = "snapshot"
+			if (ENV.SNAPSHOTS_URL) {
+				url ENV.SNAPSHOTS_URL
+				credentials {
+					username ENV.SNAPSHOTS_USERNAME
+					password ENV.SNAPSHOTS_PASSWORD
+				}
+			}
+		}
+	}
+}
+
+tasks.withType(PublishToMavenRepository) {
+	onlyIf {
+		(repository == publishing.repositories.release && publication == publishing.publications.pluginMaven) ||
+				(repository == publishing.repositories.snapshot
+				&& (publication == publishing.publications.snapshot
+				|| publication == publishing.publications.snapshotPlugin)) ||
+				(repository == publishing.repositories.release && publication.name == "quiltLoomPluginMarkerMaven")
 	}
 }
 
@@ -454,4 +481,4 @@ abstract class PrintActionsTestName extends DefaultTask {
 	}
 }
 
-apply from: rootProject.file('gradle/versions.gradle')
\ No newline at end of file
+apply from: rootProject.file('gradle/versions.gradle')
diff --git a/gradle.properties b/gradle.properties
index 31014b156127886c44921900f480e68d4528da25..ee1d4c3083cdb729df09522e88a41351d332a875 100644
--- a/gradle.properties
+++ b/gradle.properties
@@ -1,5 +1,5 @@
-name = fabric-loom
-description = The Gradle plugin for Fabric
-url = https://github.com/FabricMC/fabric-loom
+name = loom
+description = The Gradle plugin for Quilt
+url = https://github.com/QuiltMC/quilt-loom
 
 kotlin.stdlib.default.dependency = false
\ No newline at end of file
diff --git a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
index 89cdbf7a4fac78aa39bece47fc35c4e615af2594..d19cd534d53871aa1f2618cf493b66a3676a6f4c 100644
--- a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
@@ -76,7 +76,7 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
 	}
 
 	public void apply(Project project) {
-		project.getLogger().lifecycle("Fabric Loom: " + LOOM_VERSION);
+		project.getLogger().lifecycle("Quilt Loom: " + LOOM_VERSION);
 		LibraryLocationLogger.logLibraryVersions();
 
 		// Apply default plugins
diff --git a/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java b/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
index 896c72df20c7f64a9e44ec5883fadb21c1e6f5cb..743a5847aaeb36968739c8c3c39da67862ec95e5 100644
--- a/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
@@ -67,7 +67,10 @@ public class LoomRepositoryPlugin implements Plugin<PluginAware> {
 			repo.setName("Fabric");
 			repo.setUrl(MirrorUtil.getFabricRepository(target));
 		});
-
+		repositories.maven(repo -> {
+			repo.setName("Quilt Releases");
+			repo.setUrl(MirrorUtil.getQuiltRepository(target));
+		});
 		MavenArtifactRepository mojangRepo = repositories.maven(repo -> {
 			repo.setName("Mojang");
 			repo.setUrl(MirrorUtil.getLibrariesBase(target));
diff --git a/src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java b/src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java
index 2728b23537d2e36ccf130c029083f0044bd53253..4d4398c3682e8ac908df59fe881be692318db9b6 100644
--- a/src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java
+++ b/src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java
@@ -44,7 +44,7 @@ public class LoomDependencyManager {
 		sourceRemapper.remapAll();
 
 		if (extension.getInstallerData() == null) {
-			project.getLogger().info("fabric-installer.json not found in dependencies");
+			project.getLogger().warn("installer json not found in dependencies!");
 		}
 	}
 }
diff --git a/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java b/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
index 4ee3c72d800648fadd174dd72535906183744a88..f657b1e945597b9d09cd454967e3d512f54bd9cf 100644
--- a/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
+++ b/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
@@ -47,7 +47,8 @@ import net.fabricmc.loom.util.FileSystemUtil;
 import net.fabricmc.loom.util.fmj.FabricModJsonFactory;
 
 public record ArtifactMetadata(boolean isFabricMod, RemapRequirements remapRequirements, @Nullable InstallerData installerData, MixinRemapType mixinRemapType, List<String> knownIdyBsms) {
-	private static final String INSTALLER_PATH = "fabric-installer.json";
+	private static final String QUILT_INSTALLER_PATH = "quilt_installer.json";
+	private static final String FABRIC_INSTALLER_PATH = "fabric-installer.json";
 
 	public static ArtifactMetadata create(ArtifactRef artifact, String currentLoomVersion) throws IOException {
 		boolean isFabricMod;
@@ -90,11 +91,17 @@ public record ArtifactMetadata(boolean isFabricMod, RemapRequirements remapRequi
 				}
 			}
 
-			final Path installerPath = fs.getPath(INSTALLER_PATH);
+			Path installerPath = fs.getPath(QUILT_INSTALLER_PATH);
+			boolean isQuilt = true;
+
+			if (!Files.exists(installerPath)) {
+				installerPath = fs.getPath(FABRIC_INSTALLER_PATH);
+				isQuilt = false;
+			}
 
 			if (isFabricMod && Files.exists(installerPath)) {
 				final JsonObject jsonObject = LoomGradlePlugin.GSON.fromJson(Files.readString(installerPath, StandardCharsets.UTF_8), JsonObject.class);
-				installerData = new InstallerData(artifact.version(), jsonObject);
+				installerData = new InstallerData(artifact.version(), jsonObject, isQuilt);
 			}
 		}
 
diff --git a/src/main/java/net/fabricmc/loom/extension/LoomFilesBaseImpl.java b/src/main/java/net/fabricmc/loom/extension/LoomFilesBaseImpl.java
index 7dd74d5c1bcabdbf31ae1638eafca09e8d694f4d..3844b3087f8a9e4a9310434ff85400818d9ff007 100644
--- a/src/main/java/net/fabricmc/loom/extension/LoomFilesBaseImpl.java
+++ b/src/main/java/net/fabricmc/loom/extension/LoomFilesBaseImpl.java
@@ -46,22 +46,22 @@ public abstract class LoomFilesBaseImpl implements LoomFiles {
 
 	@Override
 	public File getUserCache() {
-		return createFile(getGradleUserHomeDir(), "caches" + File.separator + "fabric-loom");
+		return createFile(getGradleUserHomeDir(), "caches" + File.separator + "quilt-loom");
 	}
 
 	@Override
 	public File getRootProjectPersistentCache() {
-		return createFile(getRootDir(), ".gradle" + File.separator + "loom-cache");
+		return createFile(getRootDir(), ".gradle" + File.separator + "quilt-loom-cache");
 	}
 
 	@Override
 	public File getProjectPersistentCache() {
-		return createFile(getProjectDir(), ".gradle" + File.separator + "loom-cache");
+		return createFile(getProjectDir(), ".gradle" + File.separator + "quilt-loom-cache");
 	}
 
 	@Override
 	public File getProjectBuildCache() {
-		return createFile(getBuildDir(), "loom-cache");
+		return createFile(getBuildDir(), "quilt-loom-cache");
 	}
 
 	@Override
diff --git a/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java b/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java
index c37ded2c5418957499a211c62dd1fd0ed900dbe7..7cbed11774d38a15d4793dd6700d831f516424ba 100644
--- a/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java
+++ b/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java
@@ -129,15 +129,17 @@ public abstract class GenerateDLIConfigTask extends AbstractLoomTask {
 		}
 
 		final LaunchConfig launchConfig = new LaunchConfig()
-				.property("fabric.development", "true")
-				.property("fabric.remapClasspathFile", getRemapClasspathFile().get().getAsFile().getAbsolutePath())
+				.property("loader.development", "true")
+				.property("loader.remapClasspathFile", getRemapClasspathFile().get().getAsFile().getAbsolutePath())
 				.property("log4j.configurationFile", getLog4jConfigPaths().get())
 				.property("log4j2.formatMsgNoLookups", "true")
 
 				.argument("client", "--assetIndex")
 				.argument("client", versionInfo.assetIndex().fabricId(getMinecraftVersion().get()))
 				.argument("client", "--assetsDir")
-				.argument("client", assetsDirectory.getAbsolutePath());
+				.argument("client", assetsDirectory.getAbsolutePath())
+				.argument("client", "--version")
+				.argument("client", "Quilt Loom");
 
 		if (versionInfo.hasNativesToExtract()) {
 			String nativesPath = getNativesDirectoryPath().get();
@@ -148,12 +150,12 @@ public abstract class GenerateDLIConfigTask extends AbstractLoomTask {
 		}
 
 		if (getSplitSourceSets().get()) {
-			launchConfig.property("client", "fabric.gameJarPath.client", getClientGameJarPath().get());
-			launchConfig.property("fabric.gameJarPath", getCommonGameJarPath().get());
+			launchConfig.property("client", "loader.gameJarPath.client", getClientGameJarPath().get());
+			launchConfig.property("loader.gameJarPath", getCommonGameJarPath().get());
 		}
 
 		if (getClassPathGroups().isPresent()) {
-			launchConfig.property("fabric.classPathGroups", getClassPathGroups().get());
+			launchConfig.property("loader.classPathGroups", getClassPathGroups().get());
 		}
 
 		//Enable ansi by default for idea and vscode when gradle is not ran with plain console.
@@ -209,6 +211,11 @@ public abstract class GenerateDLIConfigTask extends AbstractLoomTask {
 		}
 
 		public LaunchConfig property(String side, String key, String value) {
+			if (key.startsWith("loader.")) {
+				// add the fabric loader property too for compat
+				property(side, "fabric." + key.substring("loader.".length()), value);
+			}
+
 			values.computeIfAbsent(side + "Properties", (s -> new ArrayList<>()))
 					.add(String.format("%s=%s", key, value));
 			return this;
diff --git a/src/main/java/net/fabricmc/loom/task/service/SourceRemapperService.java b/src/main/java/net/fabricmc/loom/task/service/SourceRemapperService.java
index 13b197f7d60fb326830b90f4b5bc176db7f827eb..617be7ffe4743c7883bc3a481dc94f6976ac1d7f 100644
--- a/src/main/java/net/fabricmc/loom/task/service/SourceRemapperService.java
+++ b/src/main/java/net/fabricmc/loom/task/service/SourceRemapperService.java
@@ -91,7 +91,7 @@ public final class SourceRemapperService extends Service<SourceRemapperService.O
 		// Create a temp directory with all of the sources
 		if (!Files.isDirectory(source)) {
 			isSrcTmp = true;
-			srcPath = Files.createTempDirectory("fabric-loom-src");
+			srcPath = Files.createTempDirectory("quilt-loom-src");
 			ZipUtils.unpackAll(source, srcPath);
 		}
 
diff --git a/src/main/java/net/fabricmc/loom/util/Constants.java b/src/main/java/net/fabricmc/loom/util/Constants.java
index b03c6a21c4eeffec0298febd31ffb5670a7cc316..be9dcfe277e57ec4b904812ff4453ec140a78637 100644
--- a/src/main/java/net/fabricmc/loom/util/Constants.java
+++ b/src/main/java/net/fabricmc/loom/util/Constants.java
@@ -33,6 +33,7 @@ public class Constants {
 	public static final String VERSION_MANIFESTS = "https://piston-meta.mojang.com/mc/game/version_manifest_v2.json";
 	public static final String EXPERIMENTAL_VERSIONS = "https://maven.fabricmc.net/net/minecraft/experimental_versions.json";
 	public static final String FABRIC_REPOSITORY = "https://maven.fabricmc.net/";
+	public static final String QUILT_REPOSITORY = "https://maven.quiltmc.org/repository/release/";
 
 	public static final int ASM_VERSION = Opcodes.ASM9;
 	public static final String RELEASE_TIME_1_3 = "2012-07-25T22:00:00+00:00";
diff --git a/src/main/java/net/fabricmc/loom/util/MirrorUtil.java b/src/main/java/net/fabricmc/loom/util/MirrorUtil.java
index 3350a985f73e403f086a2a988283cbe4644d0cc2..eef1cd42b1e53497012165e1580c6199600f2f04 100644
--- a/src/main/java/net/fabricmc/loom/util/MirrorUtil.java
+++ b/src/main/java/net/fabricmc/loom/util/MirrorUtil.java
@@ -66,4 +66,12 @@ public class MirrorUtil {
 
 		return Constants.FABRIC_REPOSITORY;
 	}
+
+	public static String getQuiltRepository(ExtensionAware aware) {
+		if (aware.getExtensions().getExtraProperties().has("loom_quilt_repository")) {
+			return String.valueOf(aware.getExtensions().getExtraProperties().get("loom_quilt_repository"));
+		}
+
+		return Constants.QUILT_REPOSITORY;
+	}
 }
diff --git a/src/main/java/net/fabricmc/loom/util/SourceRemapper.java b/src/main/java/net/fabricmc/loom/util/SourceRemapper.java
index a574cf05cb335251a98f5c71cb8ae9059bd2e86f..70f601c581462423a02990cc651a6e9e99818336 100644
--- a/src/main/java/net/fabricmc/loom/util/SourceRemapper.java
+++ b/src/main/java/net/fabricmc/loom/util/SourceRemapper.java
@@ -128,7 +128,7 @@ public class SourceRemapper {
 		if (!source.isDirectory()) {
 			// create tmp directory
 			isSrcTmp = true;
-			srcPath = Files.createTempDirectory("fabric-loom-src");
+			srcPath = Files.createTempDirectory("quilt-loom-src");
 			ZipUtils.unpackAll(source.toPath(), srcPath);
 		}
 
diff --git a/src/main/java/net/fabricmc/loom/util/gradle/GradleUtils.java b/src/main/java/net/fabricmc/loom/util/gradle/GradleUtils.java
index 6de98bff964016d9eca49c7b46dbe038a285a72e..da122564407d1b0241ad617a40d1130dd8b5f3f9 100644
--- a/src/main/java/net/fabricmc/loom/util/gradle/GradleUtils.java
+++ b/src/main/java/net/fabricmc/loom/util/gradle/GradleUtils.java
@@ -65,7 +65,7 @@ public final class GradleUtils {
 	}
 
 	public static boolean isLoomProject(Project project) {
-		return project.getPluginManager().hasPlugin("fabric-loom");
+		return project.getPluginManager().hasPlugin("org.quiltmc.loom");
 	}
 
 	public static Provider<Boolean> getBooleanPropertyProvider(Project project, String key) {
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy
index 120d9e73b50bde3e5ce3775ae9f0df1f18fadbdd..773756633e3c741c67a85bf1be7020a8f9834e66 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy
@@ -46,6 +46,7 @@ class DecompileTest extends Specification implements GradleProjectTestTrait {
 
 		where:
 		decompiler 		| task								| version
+		'vineflower'	| "genSourcesWithVineflower"		| PRE_RELEASE_GRADLE
 		'fernflower'	| "genSourcesWithFernFlower"		| PRE_RELEASE_GRADLE
 		'cfr' 			| "genSourcesWithCfr"				| PRE_RELEASE_GRADLE
 		'vineflower' 	| "genSourcesWithVineflower"		| PRE_RELEASE_GRADLE
diff --git a/src/test/groovy/net/fabricmc/loom/test/util/GradleProjectTestTrait.groovy b/src/test/groovy/net/fabricmc/loom/test/util/GradleProjectTestTrait.groovy
index b918378704862f36133799c9cb66018744cc324a..794510fa050d0cfc3287597e7d9b51e10fce0179 100644
--- a/src/test/groovy/net/fabricmc/loom/test/util/GradleProjectTestTrait.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/util/GradleProjectTestTrait.groovy
@@ -258,7 +258,7 @@ trait GradleProjectTestTrait {
 		}
 
 		File getGeneratedSources(String mappings, String jarType = "merged") {
-			return new File(getGradleHomeDir(), "caches/fabric-loom/minecraftMaven/net/minecraft/minecraft-${jarType}/${mappings}/minecraft-${jarType}-${mappings}-sources.jar")
+			return new File(getGradleHomeDir(), "caches/quilt-loom/minecraftMaven/net/minecraft/minecraft-${jarType}/${mappings}/minecraft-${jarType}-${mappings}-sources.jar")
 		}
 
 		File getGeneratedLocalSources(String mappings) {
